File: frontend/src/components/player/VideoAnalysis.js
import { useState, useEffect } from 'react';
import { api } from '@/lib/api';
import { Button } from '@/components/ui/button';
import { Progress } from '@/components/ui/progress';
import { toast } from 'sonner';
import { 
  Video, Sparkles, Target, Zap, Brain, Trophy, 
  AlertCircle, CheckCircle, Clock, RefreshCw,
  TrendingUp, Star, Activity
} from 'lucide-react';

const VideoAnalysis = () => {
  const [analysis, setAnalysis] = useState(null);
  const [status, setStatus] = useState('loading');
  const [polling, setPolling] = useState(false);

  useEffect(() => {
    loadAnalysis();
  }, []);

  useEffect(() => {
    let interval;
    if (polling) {
      interval = setInterval(() => {
        checkStatus();
      }, 5000);
    }
    return () => clearInterval(interval);
  }, [polling]);

  const loadAnalysis = async () => {
    try {
      const response = await api.getVideoAnalysis();
      if (response.data.status === 'completed') {
        setAnalysis(response.data);
        setStatus('completed');
      } else if (response.data.status === 'analyzing') {
        setStatus('analyzing');
        setPolling(true);
      } else if (response.data.status === 'failed') {
        setStatus('failed');
        setAnalysis(response.data);
      } else {
        setStatus('not_analyzed');
      }
    } catch (error) {
      setStatus('not_analyzed');
    }
  };

  const checkStatus = async () => {
    try {
      const response = await api.getVideoAnalysisStatus();
      if (response.data.status === 'completed') {
        setPolling(false);
        loadAnalysis();
      } else if (response.data.status === 'failed') {
        setPolling(false);
        setStatus('failed');
        setAnalysis(response.data);
      }
    } catch (error) {
      console.error('Status check failed');
    }
  };

  const handleTriggerAnalysis = async () => {
    try {
      await api.triggerVideoAnalysis();
      toast.success('Video analysis started! This may take a few minutes.');
      setStatus('analyzing');
      setPolling(true);
    } catch (error) {
      toast.error(error.response?.data?.detail || 'Failed to start analysis');
    }
  };

  const getScoreColor = (score) => {
    if (score >= 80) return 'text-green-500';
    if (score >= 60) return 'text-primary';
    if (score >= 40) return 'text-yellow-500';
    return 'text-red-500';
  };

  const getQualityBadge = (quality) => {
    const colors = {
      excellent: 'bg-green-500/10 text-green-500 border-green-500/20',
      good: 'bg-primary/10 text-primary border-primary/20',
      fair: 'bg-yellow-500/10 text-yellow-500 border-yellow-500/20',
      poor: 'bg-red-500/10 text-red-500 border-red-500/20',
    };
    return colors[quality?.toLowerCase()] || colors.fair;
  };

  if (status === 'loading') {
    return (
      <div className="p-8 flex items-center justify-center">
        <div className="text-primary text-xl font-heading">LOADING...</div>
      </div>
    );
  }

  if (status === 'not_analyzed') {
    return (
      <div className="p-8">
        <div className="mb-8">
          <h1 className="text-3xl font-heading font-bold uppercase mb-2">AI VIDEO ANALYSIS</h1>
          <p className="text-muted-foreground">Get AI-powered insights from your highlight video</p>
        </div>

        <div className="max-w-2xl mx-auto">
          <div className="bg-card border border-border/50 p-12 rounded-sm text-center">
            <div className="w-20 h-20 mx-auto mb-6 rounded-full bg-primary/10 flex items-center justify-center">
              <Sparkles className="w-10 h-10 text-primary" />
            </div>
            <h2 className="text-2xl font-heading font-bold uppercase mb-4">Analyze Your Highlight Video</h2>
            <p className="text-muted-foreground mb-8 max-w-md mx-auto">
              Our AI will analyze your highlight video to assess video quality, technical skills, 
              physical attributes, and identify key moments in your footage.
            </p>
            <Button
              data-testid="trigger-analysis-btn"
              onClick={handleTriggerAnalysis}
              className="bg-primary text-black font-bold uppercase tracking-wide hover:bg-primary/90 h-12 px-8"
            >
              <Video className="w-5 h-5 mr-2" />
              ANALYZE MY VIDEO
            </Button>
            <p className="text-xs text-muted-foreground mt-4">
              Make sure you have a highlight video URL set in your profile
            </p>
          </div>
        </div>
      </div>
    );
  }

  if (status === 'analyzing') {
    return (
      <div className="p-8">
        <div className="mb-8">
          <h1 className="text-3xl font-heading font-bold uppercase mb-2">AI VIDEO ANALYSIS</h1>
          <p className="text-muted-foreground">Analyzing your highlight video...</p>
        </div>

        <div className="max-w-2xl mx-auto">
          <div className="bg-card border border-border/50 p-12 rounded-sm text-center">
            <div className="w-20 h-20 mx-auto mb-6 rounded-full bg-blue-500/10 flex items-center justify-center animate-pulse">
              <RefreshCw className="w-10 h-10 text-blue-500 animate-spin" />
            </div>
            <h2 className="text-2xl font-heading font-bold uppercase mb-4">Analysis in Progress</h2>
            <p className="text-muted-foreground mb-6">
              Our AI is watching and analyzing your highlight video. This typically takes 2-5 minutes.
            </p>
            <div className="space-y-4">
              <div className="flex items-center justify-center gap-2 text-sm text-muted-foreground">
                <Clock className="w-4 h-4" />
                <span>Downloading video...</span>
              </div>
              <div className="flex items-center justify-center gap-2 text-sm text-muted-foreground">
                <Brain className="w-4 h-4" />
                <span>AI analyzing footage...</span>
              </div>
              <div className="flex items-center justify-center gap-2 text-sm text-muted-foreground">
                <Target className="w-4 h-4" />
                <span>Extracting performance data...</span>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }

  if (status === 'failed') {
    return (
      <div className="p-8">
        <div className="mb-8">
          <h1 className="text-3xl font-heading font-bold uppercase mb-2">AI VIDEO ANALYSIS</h1>
          <p className="text-muted-foreground">Analysis encountered an error</p>
        </div>

        <div className="max-w-2xl mx-auto">
          <div className="bg-card border border-red-500/50 p-12 rounded-sm text-center">
            <div className="w-20 h-20 mx-auto mb-6 rounded-full bg-red-500/10 flex items-center justify-center">
              <AlertCircle className="w-10 h-10 text-red-500" />
            </div>
            <h2 className="text-2xl font-heading font-bold uppercase mb-4">Analysis Failed</h2>
            <p className="text-muted-foreground mb-4">
              {analysis?.error || 'An error occurred during video analysis.'}
            </p>
            <p className="text-sm text-muted-foreground mb-8">
              Please ensure your video URL is valid and accessible (YouTube, Vimeo, or direct link).
            </p>
            <Button
              data-testid="retry-analysis-btn"
              onClick={handleTriggerAnalysis}
              className="bg-primary text-black font-bold uppercase tracking-wide hover:bg-primary/90 h-12 px-8"
            >
              <RefreshCw className="w-5 h-5 mr-2" />
              TRY AGAIN
            </Button>
          </div>
        </div>
      </div>
    );
  }

  // Completed analysis view
  const data = analysis?.analysis || {};

  return (
    <div className="p-8">
      <div className="mb-8 flex items-center justify-between">
        <div>
          <h1 className="text-3xl font-heading font-bold uppercase mb-2">AI VIDEO ANALYSIS</h1>
          <p className="text-muted-foreground">AI-powered assessment of your highlight video</p>
        </div>
        <Button
          data-testid="reanalyze-btn"
          variant="outline"
          onClick={handleTriggerAnalysis}
          className="border-primary text-primary hover:bg-primary hover:text-black"
        >
          <RefreshCw className="w-4 h-4 mr-2" />
          RE-ANALYZE
        </Button>
      </div>

      <div className="max-w-5xl">
        {/* Overall Score Card */}
        <div className="bg-card border border-border/50 p-8 rounded-sm mb-6">
          <div className="flex items-center justify-between">
            <div className="flex items-center gap-6">
              <div className="w-24 h-24 rounded-full bg-primary/10 flex items-center justify-center border-4 border-primary">
                <span className={`text-4xl font-heading font-bold ${getScoreColor(analysis?.overall_score || 0)}`}>
                  {analysis?.overall_score || 0}
                </span>
              </div>
              <div>
                <h2 className="text-2xl font-heading font-bold uppercase">OVERALL RATING</h2>
                <p className="text-muted-foreground">Based on AI video analysis</p>
                <div className="flex items-center gap-2 mt-2">
                  <CheckCircle className="w-4 h-4 text-green-500" />
                  <span className="text-sm text-green-500">Analysis Complete</span>
                  <span className="text-xs text-muted-foreground ml-2">
                    {analysis?.analyzed_at ? new Date(analysis.analyzed_at).toLocaleDateString() : ''}
                  </span>
                </div>
              </div>
            </div>
            <div className="text-right">
              <p className="text-sm text-muted-foreground uppercase mb-1">Recommended Level</p>
              <span className="text-xl font-heading font-bold text-primary uppercase">
                {data.recommended_level || 'N/A'}
              </span>
            </div>
          </div>
        </div>

        {/* Scout Summary */}
        {data.scout_summary && (
          <div className="bg-card border border-primary/50 p-6 rounded-sm mb-6">
            <div className="flex items-start gap-4">
              <Trophy className="w-6 h-6 text-primary flex-shrink-0 mt-1" />
              <div>
                <h3 className="font-heading font-bold uppercase mb-2">SCOUT SUMMARY</h3>
                <p className="text-muted-foreground">{data.scout_summary}</p>
                {data.similar_player_style && (
                  <p className="text-sm text-primary mt-2">
                    Playing style similar to: <span className="font-bold">{data.similar_player_style}</span>
                  </p>
                )}
              </div>
            </div>
          </div>
        )}

        <div className="grid grid-cols-1 md:grid-cols-2 gap-6 mb-6">
          {/* Video Quality */}
          <div className="bg-card border border-border/50 p-6 rounded-sm">
            <div className="flex items-center gap-3 mb-4">
              <Video className="w-5 h-5 text-blue-500" />
              <h3 className="font-heading font-bold uppercase">VIDEO QUALITY</h3>
              <span className={`ml-auto text-2xl font-bold ${getScoreColor(data.video_quality?.overall_score || 0)}`}>
                {data.video_quality?.overall_score || 0}
              </span>
            </div>
            <div className="space-y-3">
              {['resolution_quality', 'lighting', 'stability', 'clarity'].map((key) => (
                <div key={key} className="flex items-center justify-between">
                  <span className="text-sm text-muted-foreground capitalize">{key.replace('_', ' ')}</span>
                  <span className={`text-xs px-2 py-0.5 rounded-sm border ${getQualityBadge(data.video_quality?.[key])}`}>
                    {data.video_quality?.[key] || 'N/A'}
                  </span>
                </div>
              ))}
            </div>
            {data.video_quality?.notes && (
              <p className="text-xs text-muted-foreground mt-4 pt-4 border-t border-border">
                {data.video_quality.notes}
              </p>
            )}
          </div>

          {/* Player Assessment */}
          <div className="bg-card border border-border/50 p-6 rounded-sm">
            <div className="flex items-center gap-3 mb-4">
              <Target className="w-5 h-5 text-primary" />
              <h3 className="font-heading font-bold uppercase">PLAYER ASSESSMENT</h3>
              <span className={`ml-auto text-2xl font-bold ${getScoreColor(data.player_assessment?.overall_rating || 0)}`}>
                {data.player_assessment?.overall_rating || 0}
              </span>
            </div>
            <div className="space-y-3">
              <div className="flex items-center justify-between">
                <span className="text-sm text-muted-foreground">Detected Position</span>
                <span className="text-sm font-bold text-primary">{data.player_assessment?.detected_position || 'Unknown'}</span>
              </div>
              <div className="flex items-center justify-between">
                <span className="text-sm text-muted-foreground">Confidence</span>
                <span className={`text-xs px-2 py-0.5 rounded-sm border ${getQualityBadge(data.player_assessment?.confidence_level)}`}>
                  {data.player_assessment?.confidence_level || 'N/A'}
                </span>
              </div>
            </div>
            {data.player_assessment?.playing_style && (
              <p className="text-xs text-muted-foreground mt-4 pt-4 border-t border-border">
                <span className="font-medium">Style:</span> {data.player_assessment.playing_style}
              </p>
            )}
          </div>
        </div>

        {/* Technical Skills */}
        <div className="bg-card border border-border/50 p-6 rounded-sm mb-6">
          <div className="flex items-center gap-3 mb-6">
            <Zap className="w-5 h-5 text-yellow-500" />
            <h3 className="font-heading font-bold uppercase">TECHNICAL SKILLS</h3>
          </div>
          <div className="grid grid-cols-2 md:grid-cols-3 gap-4">
            {['ball_control', 'passing', 'shooting', 'dribbling', 'first_touch', 'heading'].map((skill) => {
              const value = data.technical_skills?.[skill];
              if (value === null || value === undefined) return null;
              return (
                <div key={skill} className="bg-background p-4 rounded-sm border border-border">
                  <div className="flex items-center justify-between mb-2">
                    <span className="text-sm text-muted-foreground capitalize">{skill.replace('_', ' ')}</span>
                    <span className={`text-lg font-bold ${getScoreColor(value)}`}>{value}</span>
                  </div>
                  <Progress value={value} className="h-2" />
                </div>
              );
            })}
          </div>
          {data.technical_skills?.weak_foot && (
            <div className="mt-4 pt-4 border-t border-border flex items-center gap-4">
              <span className="text-sm text-muted-foreground">Weak Foot:</span>
              <span className={`text-xs px-2 py-0.5 rounded-sm border ${getQualityBadge(data.technical_skills.weak_foot)}`}>
                {data.technical_skills.weak_foot}
              </span>
            </div>
          )}
          {data.technical_skills?.notes && (
            <p className="text-xs text-muted-foreground mt-4">{data.technical_skills.notes}</p>
          )}
        </div>

        <div className="grid grid-cols-1 md:grid-cols-2 gap-6 mb-6">
          {/* Physical Attributes */}
          <div className="bg-card border border-border/50 p-6 rounded-sm">
            <div className="flex items-center gap-3 mb-4">
              <Activity className="w-5 h-5 text-red-500" />
              <h3 className="font-heading font-bold uppercase">PHYSICAL ATTRIBUTES</h3>
            </div>
            <div className="space-y-3">
              {['pace', 'strength', 'stamina', 'agility'].map((attr) => (
                <div key={attr} className="flex items-center justify-between">
                  <span className="text-sm text-muted-foreground capitalize">{attr}</span>
                  <span className={`text-xs px-2 py-0.5 rounded-sm border ${getQualityBadge(data.physical_attributes?.[attr])}`}>
                    {data.physical_attributes?.[attr] || 'N/A'}
                  </span>
                </div>
              ))}
            </div>
            {data.physical_attributes?.notes && (
              <p className="text-xs text-muted-foreground mt-4 pt-4 border-t border-border">
                {data.physical_attributes.notes}
              </p>
            )}
          </div>

          {/* Tactical Awareness */}
          <div className="bg-card border border-border/50 p-6 rounded-sm">
            <div className="flex items-center gap-3 mb-4">
              <Brain className="w-5 h-5 text-purple-500" />
              <h3 className="font-heading font-bold uppercase">TACTICAL AWARENESS</h3>
            </div>
            <div className="space-y-3">
              <div className="flex items-center justify-between">
                <span className="text-sm text-muted-foreground">Positioning</span>
                <span className={`text-lg font-bold ${getScoreColor(data.tactical_awareness?.positioning || 0)}`}>
                  {data.tactical_awareness?.positioning || 0}
                </span>
              </div>
              <div className="flex items-center justify-between">
                <span className="text-sm text-muted-foreground">Decision Making</span>
                <span className={`text-lg font-bold ${getScoreColor(data.tactical_awareness?.decision_making || 0)}`}>
                  {data.tactical_awareness?.decision_making || 0}
                </span>
              </div>
              <div className="flex items-center justify-between">
                <span className="text-sm text-muted-foreground">Work Rate</span>
                <span className={`text-xs px-2 py-0.5 rounded-sm border ${getQualityBadge(data.tactical_awareness?.work_rate)}`}>
                  {data.tactical_awareness?.work_rate || 'N/A'}
                </span>
              </div>
            </div>
            {data.tactical_awareness?.notes && (
              <p className="text-xs text-muted-foreground mt-4 pt-4 border-t border-border">
                {data.tactical_awareness.notes}
              </p>
            )}
          </div>
        </div>

        {/* Strengths & Areas for Improvement */}
        <div className="grid grid-cols-1 md:grid-cols-2 gap-6 mb-6">
          {data.strengths?.length > 0 && (
            <div className="bg-card border border-green-500/50 p-6 rounded-sm">
              <div className="flex items-center gap-3 mb-4">
                <TrendingUp className="w-5 h-5 text-green-500" />
                <h3 className="font-heading font-bold uppercase">STRENGTHS</h3>
              </div>
              <ul className="space-y-2">
                {data.strengths.map((strength, i) => (
                  <li key={i} className="flex items-start gap-2 text-sm">
                    <Star className="w-4 h-4 text-green-500 flex-shrink-0 mt-0.5" />
                    <span>{strength}</span>
                  </li>
                ))}
              </ul>
            </div>
          )}

          {data.areas_for_improvement?.length > 0 && (
            <div className="bg-card border border-yellow-500/50 p-6 rounded-sm">
              <div className="flex items-center gap-3 mb-4">
                <Target className="w-5 h-5 text-yellow-500" />
                <h3 className="font-heading font-bold uppercase">AREAS TO IMPROVE</h3>
              </div>
              <ul className="space-y-2">
                {data.areas_for_improvement.map((area, i) => (
                  <li key={i} className="flex items-start gap-2 text-sm text-muted-foreground">
                    <AlertCircle className="w-4 h-4 text-yellow-500 flex-shrink-0 mt-0.5" />
                    <span>{area}</span>
                  </li>
                ))}
              </ul>
            </div>
          )}
        </div>

        {/* Key Moments */}
        {data.key_moments?.length > 0 && (
          <div className="bg-card border border-border/50 p-6 rounded-sm">
            <div className="flex items-center gap-3 mb-4">
              <Sparkles className="w-5 h-5 text-primary" />
              <h3 className="font-heading font-bold uppercase">KEY MOMENTS DETECTED</h3>
              <span className="text-sm text-muted-foreground ml-auto">{data.key_moments.length} moments</span>
            </div>
            <div className="space-y-3">
              {data.key_moments.map((moment, i) => (
                <div key={i} className="flex items-center gap-4 p-3 bg-background rounded-sm border border-border">
                  <div className="flex-shrink-0">
                    <span className={`text-xs px-2 py-1 rounded-sm uppercase font-bold ${
                      moment.type === 'goal' ? 'bg-green-500/10 text-green-500' :
                      moment.type === 'assist' ? 'bg-blue-500/10 text-blue-500' :
                      moment.type === 'skill' ? 'bg-purple-500/10 text-purple-500' :
                      'bg-primary/10 text-primary'
                    }`}>
                      {moment.type}
                    </span>
                  </div>
                  <div className="flex-1">
                    <p className="text-sm">{moment.description}</p>
                    {moment.timestamp && (
                      <p className="text-xs text-muted-foreground mt-1">{moment.timestamp}</p>
                    )}
                  </div>
                  <div className="flex items-center gap-1">
                    <Star className="w-4 h-4 text-yellow-500" />
                    <span className="font-bold">{moment.quality_rating}/10</span>
                  </div>
                </div>
              ))}
            </div>
          </div>
        )}
      </div>
    </div>
  );
};

export default VideoAnalysis;

File: backend/video_analysis.py
"""
Video Analysis Service using Google Gemini
Analyzes highlight videos for quality and performance metrics
"""

import os
import re
import json
import uuid
import asyncio
import tempfile
import time
from datetime import datetime, timezone
from typing import Optional, Dict, Any
import httpx
import google.generativeai as genai

GEMINI_API_KEY = os.environ.get("GEMINI_API_KEY", "")

ANALYSIS_PROMPT = """You are an expert football (soccer) scout and video analyst. Analyze this highlight video and provide a comprehensive assessment.

Return your analysis as a valid JSON object with the following structure:

{
  "video_quality": {
    "overall_score": <1-100>,
    "resolution_quality": "<poor/fair/good/excellent>",
    "lighting": "<poor/fair/good/excellent>",
    "stability": "<poor/fair/good/excellent>",
    "clarity": "<poor/fair/good/excellent>",
    "audio_quality": "<poor/fair/good/excellent/no_audio>",
    "notes": "<brief notes about video quality>"
  },
  "player_assessment": {
    "overall_rating": <1-100>,
    "detected_position": "<GK/CB/LB/RB/DM/CM/AM/Winger/Striker/Unknown>",
    "confidence_level": "<low/medium/high>",
    "playing_style": "<brief description of playing style>"
  },
  "technical_skills": {
    "ball_control": <1-100>,
    "passing": <1-100>,
    "shooting": <1-100>,
    "dribbling": <1-100>,
    "heading": <1-100>,
    "first_touch": <1-100>,
    "weak_foot": "<poor/fair/good/excellent/not_shown>",
    "notes": "<brief notes on technical abilities>"
  },
  "physical_attributes": {
    "pace": "<slow/average/fast/explosive/not_assessable>",
    "strength": "<weak/average/strong/dominant/not_assessable>",
    "stamina": "<low/average/high/excellent/not_assessable>",
    "agility": "<low/average/high/excellent/not_assessable>",
    "notes": "<brief notes on physical attributes>"
  },
  "tactical_awareness": {
    "positioning": <1-100>,
    "decision_making": <1-100>,
    "work_rate": "<low/average/high/excellent>",
    "defensive_contribution": "<poor/fair/good/excellent/not_applicable>",
    "notes": "<brief notes on tactical intelligence>"
  },
  "key_moments": [
    {
      "timestamp": "<approximate timestamp or description>",
      "type": "<goal/assist/tackle/save/dribble/pass/skill/other>",
      "description": "<brief description of the moment>",
      "quality_rating": <1-10>
    }
  ],
  "strengths": ["<strength 1>", "<strength 2>", "<strength 3>"],
  "areas_for_improvement": ["<area 1>", "<area 2>"],
  "scout_summary": "<2-3 sentence summary a scout would write>",
  "recommended_level": "<amateur/semi-pro/professional/elite>",
  "similar_player_style": "<name of a professional player with similar style, if applicable>"
}

Be thorough but realistic. If you cannot assess something from the video, indicate that clearly. Focus on what is actually visible in the footage."""


async def download_video_from_url(url: str, output_path: str) -> bool:
    """Download video from URL (supports YouTube, Vimeo, direct links)"""
    try:
        if any(ext in url.lower() for ext in [".mp4", ".webm", ".mov", ".avi"]):
            async with httpx.AsyncClient(timeout=120.0, follow_redirects=True) as client:
                response = await client.get(url)
                if response.status_code == 200:
                    with open(output_path, "wb") as f:
                        f.write(response.content)
                    return True
            return False

        youtube_patterns = [r"(?:youtube\.com/watch\?v=|youtu\.be/|youtube\.com/embed/)([a-zA-Z0-9_-]+)"]
        vimeo_pattern = r"vimeo\.com/(\d+)"

        is_youtube = any(re.search(p, url) for p in youtube_patterns)
        is_vimeo = re.search(vimeo_pattern, url)

        if is_youtube or is_vimeo:
            cmd = ["yt-dlp", "-f", "best", "-o", output_path, "--no-playlist", "--socket-timeout", "60", "--max-filesize", "50m", "--no-check-certificates", url]
            process = await asyncio.create_subprocess_exec(*cmd, stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE)
            stdout, stderr = await asyncio.wait_for(process.communicate(), timeout=180)
            if process.returncode == 0 and os.path.exists(output_path):
                return True
            print(f"yt-dlp failed: {stderr.decode()}")
            return False

        async with httpx.AsyncClient(timeout=120.0, follow_redirects=True) as client:
            response = await client.get(url)
            if response.status_code == 200:
                content_type = response.headers.get("content-type", "")
                if "video" in content_type or len(response.content) > 100000:
                    with open(output_path, "wb") as f:
                        f.write(response.content)
                    return True

        return False

    except Exception as e:
        print(f"Error downloading video: {e}")
        return False


async def analyze_video_with_gemini(video_path: str) -> Dict[str, Any]:
    """Analyze video using Google Gemini Files API"""
    try:
        genai.configure(api_key=GEMINI_API_KEY)

        print("Uploading video to Gemini Files API...")
        video_file = genai.upload_file(path=video_path, mime_type="video/mp4")

        max_wait = 120
        waited = 0
        while video_file.state.name == "PROCESSING" and waited < max_wait:
            time.sleep(3)
            waited += 3
            video_file = genai.get_file(video_file.name)

        if video_file.state.name == "FAILED":
            return {"success": False, "error": "Video processing failed in Gemini Files API"}

        model = genai.GenerativeModel(
            model_name="gemini-2.0-flash",
            system_instruction="You are an expert football scout and video analyst. Always respond with valid JSON only, no additional text."
        )

        response = model.generate_content([video_file, ANALYSIS_PROMPT])

        try:
            genai.delete_file(video_file.name)
        except Exception:
            pass

        response_text = response.text.strip()
        if response_text.startswith("```json"):
            response_text = response_text[7:]
        if response_text.startswith("```"):
            response_text = response_text[3:]
        if response_text.endswith("```"):
            response_text = response_text[:-3]

        analysis = json.loads(response_text.strip())
        return {"success": True, "analysis": analysis}

    except json.JSONDecodeError as e:
        print(f"JSON parse error: {e}")
        return {"success": False, "error": "Failed to parse AI response"}
    except Exception as e:
        print(f"Gemini analysis error: {e}")
        return {"success": False, "error": str(e)}


async def analyze_highlight_video(video_url: str, player_id: str) -> Dict[str, Any]:
    """Main function to analyze a highlight video"""
    analysis_id = str(uuid.uuid4())
    temp_dir = tempfile.mkdtemp()
    video_path = os.path.join(temp_dir, f"video_{analysis_id}.mp4")

    try:
        print(f"Downloading video from {video_url}...")
        download_success = await download_video_from_url(video_url, video_path)

        if not download_success or not os.path.exists(video_path):
            return {"success": False, "error": "Failed to download video. Please ensure the URL is valid and accessible.", "analysis_id": analysis_id}

        file_size = os.path.getsize(video_path)
        if file_size > 50 * 1024 * 1024:
            return {"success": False, "error": "Video file is too large (>50MB). Please use a shorter highlight clip.", "analysis_id": analysis_id}

        print(f"Video downloaded: {file_size / 1024 / 1024:.2f}MB")
        print("Analyzing video with AI...")
        result = await analyze_video_with_gemini(video_path)

        if result["success"]:
            return {
                "success": True,
                "analysis_id": analysis_id,
                "player_id": player_id,
                "video_url": video_url,
                "analysis": result["analysis"],
                "analyzed_at": datetime.now(timezone.utc).isoformat()
            }
        else:
            return {"success": False, "error": result.get("error", "Analysis failed"), "analysis_id": analysis_id}

    except Exception as e:
        print(f"Video analysis error: {e}")
        return {"success": False, "error": str(e), "analysis_id": analysis_id}
    finally:
        try:
            if os.path.exists(video_path):
                os.remove(video_path)
            os.rmdir(temp_dir)
        except Exception:
            pass


def calculate_overall_score(analysis: Dict[str, Any]) -> int:
    """Calculate overall player score from analysis"""
    try:
        scores = []

        if "video_quality" in analysis:
            scores.append(analysis["video_quality"].get("overall_score", 50) * 0.1)

        if "player_assessment" in analysis:
            scores.append(analysis["player_assessment"].get("overall_rating", 50) * 0.2)

        if "technical_skills" in analysis:
            tech = analysis["technical_skills"]
            tech_scores = [tech.get("ball_control", 50), tech.get("passing", 50), tech.get("shooting", 50), tech.get("dribbling", 50), tech.get("first_touch", 50)]
            scores.append(sum(tech_scores) / len(tech_scores) * 0.35)

        if "tactical_awareness" in analysis:
            tact = analysis["tactical_awareness"]
            tact_scores = [tact.get("positioning", 50), tact.get("decision_making", 50)]
            scores.append(sum(tact_scores) / len(tact_scores) * 0.25)

        if "key_moments" in analysis:
            moments = analysis["key_moments"]
            if moments:
                avg_moment_quality = sum(m.get("quality_rating", 5) for m in moments) / len(moments)
                scores.append(avg_moment_quality * 10 * 0.1)

        return min(100, max(0, int(sum(scores))))

    except Exception as e:
        print(f"Score calculation error: {e}")
        return 50
